client/lib/sites-list/list.js

```diff
@@ -167,28 +167,44 @@ SitesList.prototype.createSiteObject = function( site ) {
 		return Site( site );
 	}
 };
+
 /**
  * Marks collisions between .com sites and Jetpack sites that have the same URL
  * Add the hasConflict attribute to .com sites that collide with Jetpack sites.
  *
  * @api private
  *
+ * @param {Array} sites list of stored sites
  */
 SitesList.prototype.markCollisions = function( sites ) {
-	sites.forEach( function( site, index, collisions ) {
-		var hasCollision;
+	const jetpackUrls = new Map();
+
+	const addJetpackSiteUrl = ( { ID, URL } ) => {
+		const baseUrl = withoutHttp( URL );
+		let siteSet = jetpackUrls.get( baseUrl );
+
+		if ( undefined === siteSet ) {
+			siteSet = new Set();
+			jetpackUrls.set( baseUrl, siteSet );
+		}
+
+		siteSet.add( ID );
+	};
+
+	// first make sure all Jetpack sites are accounted for
+	sites.forEach( site => site.jetpack && addJetpackSiteUrl( site ) );
 
+	// then find non-Jetpack sites sharing the same URL
+	sites.forEach( site => {
 		if ( ! site.jetpack ) {
-			hasCollision = some( collisions, function( someSite ) {
-				return (
-					someSite.jetpack &&
-					site.ID !== someSite.ID &&
-					withoutHttp( site.URL ) === withoutHttp( someSite.URL )
-				);
-			} );
-			if ( hasCollision ) {
-				site.hasConflict = true;
-			}
+			return;
+		}
+
+		const jetpackSites = jetpackUrls.get( withoutHttp( site.URL ) );
+		const siteCount = jetpackSites.size;
+
+		if ( siteCount > 1 || ( siteCount === 1 || ! jetpackSites.has( site.ID ) ) ) {
```

Comment:
I think we should be able to just test for `siteCount > 0` here. No?
Since we are looking for jetpack sites that have the same url.

Reply:
Mainly I was trying to update the code without changing the behavior. The purpose of this function was unclear to me and I wasn't sure if we'd end up with sites sharing the same ids as the corresponding Jetpack sites. Therefore, I wrote it this way because if we have more than one site tracked we know there's another site with the same ID that isn't the one under inspection.

Honestly some context would be really helpful here! What is this function supposed to do, what example inputs would we expect to mark as collisions? what possibilities could we get here that we need to handle?